proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+# Background
+
+Cryptography is an essential part of modern computing and information security, but it requires care, nuance, and knowledge to properly implement, and is extremely easy to mess up. It covers a diverse range of functions, with individual algorithms often having their own particular quirks that must be acknowledged to properly and safely use. Failure to properly implement cryptography carries the risk of severely compromising the security of any system in which it is used. Manually implementing common cryptographic primitives such as hashes, ciphers, digital signatures, big-integer arithmetic is thus considered extremely unwise, both in terms of security and performance.
+
+Cryptography is even harder for functional programming languages. Properly implementing techniques such as sanitizing memory, performing operations in constant time, are already difficult enough in languages such as C, and in a lazy functional language it is made even more difficult by such things as the potential for references to be kept alive within thunks. As a result, the space of functional cryptography is relatively under-explored and under-developed, having favored other languages where such things are easier to manage.
+
+However, if we accept this downside as a part of a trade-off, and explore the ways that functional programming is beneficial to cryptography (monads, type systems, referential transparency), there is then also the potential for functional cryptography to be made *easier*. The functional machinery of Haskell is well-suited for adding important contextual control to cryptography, and can help prevent or eliminate many issues and errors in ways that languages such as C cannot. I believe that there is room for significant improvement, given a directed effort to develop the space.
+
+# Problem Statement
+
+Cryptography in Haskell lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and exposes end users to significant risk in the event of a lapse in security. Some companies have built their own solution to this, but there is no community-driven, community-owned solution.
+
+Cryptography in Haskell is also fragile and outdated, having seen considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. This has placed the long-term stability of important libraries such as `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented.
+
+There are several reasons for this:
+
+1) Cryptography involves a lot of stateful low-level bit twiddling and random IO, something Haskell is not known for being good at. Classes like `Bits` and `FiniteBits` classes are critical for cryptography, but are terribly awkward to implement (like `Num`). Also, `ByteString` does not implement them for nuanced reasons.
```

Comment:
This is random IO in the sense of "random access to mutable arrays" rather than "IO happens sometimes", right?

A link to the ByteString issue here would be convenient.

Reply:
I should rephrase this as cryptographic non-determinism rather than 'random IO', and differentiate it from 'worldly' non-determinism, which is your more traditional IO. This is probably way more of an answer than you expected, but there is also a very fine distinction between entropy and pseudorandom generators, and existing Haskell libraries do not do a very good job of separating it all out, especially when we use pseudorandom generators as a substitute for entropy sources.

Ideally, non-deterministic values should just be exposed as an input, and the resulting function would be pure and deterministic. However, nonce mismanagement has caused *many* security lapses, and so many functions / protocols / libraries remove user randomness / nonce handling in favor of hiding it as an implementation detail, either taking a random generator state as an argument, or even hiding it as an internal global / IO reference, which makes the resulting function non-deterministic / impure. A good example of this is `bcrypt`, which takes a random generator context as an argument.

Properly representing the difference between determinism, ideal entropy, and pseudorandom generators requires differentiating functions of the following forms, even though they are all related:


```haskell
deterministicFoo :: Int -> a -> b
randomFoo :: (Random m) => a -> m b
pseudoRandomFoo :: Pseudorandom g => g -> a -> (g, b)
impureFoo :: a -> IO b
```

Note that these are not the same when it comes to cryptography and entropy. `deterministicFoo` represents a single sampling event, `randomFoo` represents an event that hasn't been sampled yet and will be different every time you do, and `pseudoRandomFoo` represents an event that has been sampled so its predetermined but you don't know the result yet. You can simulate Random with a secret state a la `(State g, Pseudorandom g) => ...` but there is a lawful difference between them.

For reference, the Botan FFI RNG objects are closer to entropy sources (although they internally use PRNG with automatic reseeding from system entropy sources).

---

Regarding `Bits` and `FiniteBits`, this is worth getting into and I made a very rough sketch of a `Boolean` class that may have spawned a conversation or two [in the devlog](https://discourse.haskell.org/t/botan-bindings-devlog/6855/78) and [in a side thread](https://discourse.haskell.org/t/why-doesnt-uuid-have-a-bits-instance/7629). Essentially, `Bits` (and the fixed-width `FiniteBits`) is a class that combines two concepts:

1) Boolean / Heyting algebra, for which there is no concept of any 'internal bits' or structure and you are just operating on some object as a whole
2) Things that are contructed from / encoded into a set of 'internal bits' using some encoding, for which individual bits are indexable / addressible, and thus individually boolean-operable. There is also a link here to indexing / representable, but that is getting way out of scope.

Notably, [the documentation itself](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Bits.html) further states that "The Bits class defines bitwise operations over integral types.". Technically this means that only things that are integer numbers should be `Bits` and `FiniteBits`, but Godel numbering rears its ugly head, and that just adds the question of whether `Bits`'s boolean operations are intended to apply to an object or its encoding, and as a result `Bits` and `FiniteBits` restricts it to objects for which the boolean operations are isomorphic to boolean operations over its encoding.

Given this understanding, `ByteString` does not have a canonical `Bits` instance because it has no canonical integer representation - a given bytestring could represent many such integers, whether it be big endian or little endian, it could have a sign bit, or be two's complement. Its bits might not be contiguous or even ordered. It could easily satisfy a `Boolean` instance, but the issue of indexing individual bits and integral type requirement interferes with a full instance of `Bits`.

> As if that wasn't confusing enough, we could go further and involve [finite](https://en.wikipedia.org/wiki/Finite_field) [fields](https://en.wikipedia.org/wiki/Field_(mathematics)) and [bit-field](https://en.wikipedia.org/wiki/Bit_field) and how bit-fields are fields of bits but not all fields of bits are bit-fields. Whether or not we should tie `Bits` and `FiniteBits` to `Field` and `FiniteField` is a whole extra conversation.